### PR TITLE
Update legacy indexes to preserve documents sort fields on update

### PR DESCRIPTION
**Pull request by MishaDemianenko**
Legacy indexes use stored fields to store user data and sorted doc value typed fields to make sorting smooth.
After retrieval of any saved document from index store only stored fields are visible (not sort fields)
and in case if we will re-save that document we will loose any sorted fields that we had before in the index for that document.
This PR introduce functionality that will restore all 'missing' sort fields to make sure that documents can be always sorted.
Also as result of fix we need to migrate potentially affected old indexes and restore potentially missing sorting fields. To be able to that legacy index store migrators updated and both standard and high level format versions upgraded to be able to use standard migration mechanism.

### CODE REVIEW

**Comment by tinwelint** [comment]
It looks like you've bumped the store format and duplicated all those formats, why did you do that? The formats are all the same, aren't they?

**Comment by tinwelint** [comment]
@MishaDemianenko I'd like it if you would squasch the two last commits, just so much more noise in a PR making a lot of changes and having another commit undoing those.

**Comment by MishaDemianenko** [comment]
@tinwelint done

File: community/lucene-index/src/main/java/org/neo4j/index/impl/lucene/legacy/IndexType.java [hunk 1]
```diff
@@ -403,25 +392,74 @@ private void clearDocument( Document document )
...
+                addNewFieldToDocument( document, key, existingValue );
+            }
+        }
+    }
+
+    private boolean haveSortField( IndexableField[] fields )
+    {
+        for ( IndexableField field : fields )
+        {
+            if ( !DocValuesType.NONE.equals( field.fieldType().docValuesType() ) )
```
Comment (tinwelint) [inline community/lucene-index/src/main/java/org/neo4j/index/impl/lucene/legacy/IndexType.java:438]: use the other `haveSortedField` method here?

File: community/consistency-check/src/test/java/org/neo4j/unsafe/impl/batchimport/ParallelBatchImporterTest.java [hunk 2]
```diff
@@ -237,7 +236,7 @@ private void assertConsistent( File storeDir ) throws ConsistencyCheckIncomplete
 
     protected String getFormatName()
     {
-        return StandardV3_0.NAME;
+        return StandardV3_0_7.NAME;
```
Comment (tinwelint) [inline community/consistency-check/src/test/java/org/neo4j/unsafe/impl/batchimport/ParallelBatchImporterTest.java:239]: (This comment applies to all similar changes): should we have a notion of "latest" standard format to not have to make wide-spread changes like this and potentially forget some?

File: community/kernel/src/main/java/org/neo4j/kernel/spi/legacyindex/IndexImplementation.java [hunk 3]
```diff
@@ -107,4 +107,11 @@
      */
     @Override
     void shutdown() throws Throwable;
+
+    /**
+     * Provide store migrator for particular index implementation that participate in store migration and
+     * should be able to perform cross version migration of index implementation
+     * @return index implementation store migration
+     */
+    StoreMigrationParticipant getStoreMigrator();
```
Comment (tinwelint) [inline community/kernel/src/main/java/org/neo4j/kernel/spi/legacyindex/IndexImplementation.java]: Provide default implementation `NOT_PARTICIPATING` to not cause problems for actual custom implementations out there?

File: enterprise/kernel/src/main/java/org/neo4j/kernel/impl/store/format/highlimit/v306/HighLimitFactoryV3_0_6.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.kernel.impl.store.format.highlimit.v306;
+
+import org.neo4j.helpers.Service;
+import org.neo4j.kernel.impl.store.format.RecordFormats;
+
+@Service.Implementation( RecordFormats.Factory.class )
+public class HighLimitFactoryV3_0_6 extends RecordFormats.Factory
```
Comment (tinwelint) [inline enterprise/kernel/src/main/java/org/neo4j/kernel/impl/store/format/highlimit/v306/HighLimitFactoryV3_0_6.java:26]: Shouldn't this be V3_0_7?

Comment (MishaDemianenko) [reply enterprise/kernel/src/main/java/org/neo4j/kernel/impl/store/format/highlimit/v306/HighLimitFactoryV3_0_6.java:26]: 3.0.6 it's a version where format was last used, while new one is always HighLimit and factory for it is HighLimitFactory. Which is different how it is for standard record format where we use version where it was introduced as part of versioning schema

File: community/consistency-check/src/test/java/org/neo4j/unsafe/impl/batchimport/ParallelBatchImporterTest.java (hunk 2 above)
Comment (MishaDemianenko) [reply community/consistency-check/src/test/java/org/neo4j/unsafe/impl/batchimport/ParallelBatchImporterTest.java:239]: yes that sounds like a good idea, card created

**Comment by MishaDemianenko** [comment]
@tinwelint updated

**Comment by MishaDemianenko** [comment]
Replaced by #8462 closing

**Pull request closed without merging**